pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,425 @@
+					return baseValue, nil
+				}
+				if !m.base.Next() || innerRange.ID != baseRange.ID {
+					break
+				}
+			}
+		} else if !m.base.NextRange() {
+			break
+		}
+	}
+	if err := m.base.Err(); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+// writeRange writes Range using writer
+func (m *merger) writeRange(writeRange *Range) error {
+	if m.logger.IsTracing() {
+		m.logger.WithFields(logging.Fields{
+			"from": string(writeRange.MinKey),
+			"to":   string(writeRange.MaxKey),
+			"ID":   writeRange.ID,
+		}).Trace("copy entire range")
+	}
+	if err := m.writer.WriteRange(*writeRange); err != nil {
+		return fmt.Errorf("copy range %s: %w", writeRange.ID, err)
+	}
+	return nil
+}
+
+// writeRecord writes graveler.ValueRecord using writer
+func (m *merger) writeRecord(writeValue *graveler.ValueRecord) error {
+	if m.logger.IsTracing() {
+		m.logger.WithFields(logging.Fields{
+			"key": string(writeValue.Key),
+			"ID":  string(writeValue.Identity),
+		}).Trace("write record")
+	}
+	if err := m.writer.WriteRecord(*writeValue); err != nil {
+		return fmt.Errorf("write record: %w", err)
+	}
+	return nil
+}
+
+// applyAll applies all remaining changes from Iterator to writer
+func (m *merger) applyAll(iter Iterator) error {
+	for {
+		select {
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		default:
+		}
+		iterValue, iterRange := iter.Value()
+		if iterValue == nil {
+			baseRange, err := m.baseRangeGE(graveler.Key(iterRange.MinKey))
+			if err != nil {
+				return fmt.Errorf("base range GE: %w", err)
+			}
+			if baseRange == nil || baseRange.ID != iterRange.ID {
+				if err := m.writeRange(iterRange); err != nil {
+					return err
+				}
+			}
+			if !iter.NextRange() {
+				break
+			}
+		} else {
+			baseValue, err := m.baseKeyGE(iterValue.Key)
+			if err != nil {
+				return fmt.Errorf("base value GE: %w", err)
+			}
+			if baseValue == nil || !bytes.Equal(baseValue.Identity, iterValue.Identity) {
+				if err := m.writeRecord(iterValue); err != nil {
+					return err
+				}
+			}
+			if !iter.Next() {
+				break
+			}
+		}
+	}
+	return iter.Err()
+}
+
+// applyBothRanges applies merging when both source and dest on range header
+func (m *merger) applyBothRanges(sourceRange *Range, destRange *Range) error {
+	switch {
+	case sourceRange.ID == destRange.ID: // source and dest added the same range
+		err := m.writeRange(sourceRange)
+		if err != nil {
+			return err
+		}
+		m.haveSource = m.source.NextRange()
+		m.haveDest = m.dest.NextRange()
+
+	case bytes.Compare(sourceRange.MaxKey, destRange.MinKey) < 0: // source before dest
+		baseRange, err := m.baseRangeGE(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest deleted this range
+			m.haveSource = m.source.NextRange()
+			return nil
+		}
+		if baseRange == nil || destRange.ID == baseRange.ID { // source added this range
+			err = m.writeRange(sourceRange)
```

Comment:
I think there's a bug here, adding a test case which I think will reflect it (add this to `Test_merge`).
I hope this is a valid case.
@guy-har, would love your input here as well.
```
		"is it a bug?": {
			baseRange: []testRange{
				{
					rng:     committed.Range{ID: "base:b-c", MinKey: committed.Key("b"), MaxKey: committed.Key("c"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "b", identity: "b"}, {key: "c", identity: "c"}},
				},
				{
					rng:     committed.Range{ID: "base:d-e", MinKey: committed.Key("d"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "d", identity: "d"}, {key: "e", identity: "e"}},
				},
			},
			sourceRange: []testRange{
				{
					rng:     committed.Range{ID: "source:a-d", MinKey: committed.Key("a"), MaxKey: committed.Key("d"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "a", identity: "a"}, {key: "b", identity: "b"}, {key: "d", identity: "d"}},
				},
				{
					rng:     committed.Range{ID: "source:e", MinKey: committed.Key("e"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "e", identity: "e1"}},
				},
			},
			destRange: []testRange{
				{
					rng:     committed.Range{ID: "base:b-c", MinKey: committed.Key("b"), MaxKey: committed.Key("c"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "b", identity: "b"}, {key: "c", identity: "c"}},
				},
				{
					rng:     committed.Range{ID: "dest:d-e", MinKey: committed.Key("d"), MaxKey: committed.Key("e"), Count: 2, EstimatedSize: 1024},
					records: []testValueRecord{{key: "d", identity: "d1"}, {key: "e", identity: "e"}},
				},
			},
			conflictExpectedIdx: nil,
			expectedActions: []writeAction{
				{
					action:   actionTypeWriteRecord,
					key:      "a",
					identity: "source:a-d",
				},
				{
					action:   actionTypeWriteRecord,
					key:      "b",
					identity: "b",
				},
				{
					action:   actionTypeWriteRecord,
					key:      "d",
					identity: "d1",
				},
				{
					action:   actionTypeWriteRecord,
					key:      "e",
					identity: "e1",
				},
			},
		}
```

Reply:
I fixed the bug and added this test, thanks!